inc/saiprefixcompression.h

```diff
@@ -0,0 +1,305 @@
+/**
+ * Copyright (c) 2024 Microsoft Open Technologies, Inc.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License"); you may
+ *    not use this file except in compliance with the License. You may obtain
+ *    a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR
+ *    CONDITIONS OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT
+ *    LIMITATION ANY IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS
+ *    FOR A PARTICULAR PURPOSE, MERCHANTABILITY OR NON-INFRINGEMENT.
+ *
+ *    See the Apache Version 2.0 License for specific language governing
+ *    permissions and limitations under the License.
+ *
+ *    Microsoft would like to thank the following companies for their review and
+ *    assistance with these files: Intel Corporation, Mellanox Technologies Ltd,
+ *    Dell Products, L.P., Facebook, Inc., Marvell International Ltd.
+ *
+ * @file    saiprefixcompression.h
+ *
+ * @brief   This module defines SAI prefix compression interface
+ */
+
+#if !defined (__SAIPREFIXCOMPRESSION_H_)
+#define __SAIPREFIXCOMPRESSION_H_
+
+#include <saitypes.h>
+
+/**
+ * @defgroup SAIPREFIXCOMPRESSION SAI - Prefix Compression API definitions
+ *
+ * @{
+ */
+
+/**
+ * @brief Attribute Id for SAI prefix compression object
+ */
+typedef enum _sai_prefix_compression_table_attr_t
+{
+    /**
+     * @brief Start of attributes
+     */
+    SAI_PREFIX_COMPRESSION_TABLE_ATTR_START,
+
+    /**
+     * @brief Label attribute used to unique identify Table.
+     *
+     * @type char
+     * @flags CREATE_AND_SET
+     * @default ""
+     */
+    SAI_PREFIX_COMPRESSION_TABLE_ATTR_LABEL = SAI_PREFIX_COMPRESSION_TABLE_ATTR_START,
+
+    /**
+     * @brief Prefix Compression table default entry BINCODE.
+     * BINCODE applied to any address that does not match a prefix in the table
+     *
+     * @type sai_uint32_t
+     * @flags MANDATORY_ON_CREATE | CREATE_AND_SET
+     */
+    SAI_PREFIX_COMPRESSION_TABLE_ATTR_DEFAULT_ENTRY_BINCODE,
+
+    /**
+     * @brief End of attributes
```

Comment:
Need the below two attributes:
1) SAI_PREFIX_COMPRESSION_TABLE_ATTR_TYPE to distinguish between Source IP and Destination IP address
2) SAI_PREFIX_COMPRESSION_TABLE_ATTR_ACL_STAGE  of type sai_acl_stage_t

Reply:
For the comment below:
1. We should not have indicate an attribute for source ip or destination IP. From creating the actual prefix tables, they should not be labeled on creation. The user can give them a unique label. The user can decide to mix source/destination values. In ACL the distinction is made because at that point we would need to know if we are matching on source address/destination. However, during table creation there is no value to add that distinction.

I think it is important to seperate the feature Prefix compression from other features that use it. As a prefix compression feature, the main purpose of it is to create tables of prefixes based on longest prefix match. There are no rules on how you create these tables. The issue comes in which feature uses these tables. If we are using these tables with ACL, then ACL would need to know which one is Source IP/DST IP and stage. ACL as a feature has these requirements. Another feature in the future that might use prefix tables, might not care.  

3. There could be value to this, currently when you apply a table to ACL you would indicate the stage, however, I am open to discussing this next time we can bring up this topic in a meeting.

Reply:
As discussed in meeting today:
1. Will wait to see if @rck-innovium has a use case that will require this
2. Will add the stage to the table

Reply:
@rck-innovium I added stage to table creation (ingress/egress), please close if you are ok with the changes

Reply:
I am ok with using a new sai_prefix_compression_stage_t instead of sai_acl_stage_t. Prefix compression is generic and can be used by features other than ACL.

I still believe we need SAI_PREFIX_COMPRESSION_TABLE_ATTR_TYPE with values Source, Destination and Both. This lets the SAI application query the enum capability and use accordingly.

Reply:
Is there a use case in your mind that would require this? since the feature that would be using prefix compression would indicate which table type it would expect. it would be up to the user to pass in the correct information. 

I don't mind adding this as a topic that we need to revisit and discuss

Reply:
Simple use-case is when the grouping of IP addresses is different for Source IP vs Dest IP. A quick example below where DIP grouping is 10.1.X.X, and SIP grouping is different for 10.1.1.X from 10.1.2.X.

| SIP       | DIP       | L4-Dst-Port | Syn  | Action |  
|-----------|-----------|-------------|------|--------|  
| \*         | 10.1.X.X  | 80          | True | Permit |  
| \*         | 10.1.X.X  | 90          | True | Permit |  
| \*         | 10.1.X.X  | \*           | True | Drop   |  
| 10.1.1.X  | 100.X.X.X | 80          | True | Allow  |  
| 10.1.1.X  | 100.X.X.X | 90          | True | Allow  |  
| 10.1.2.X  | 200.X.X.X | 80          | True | Allow  |  
| 10.1.2.X  | 200.X.X.X | 90          | True | Allow  |  
 

There are multiple possible implementations for assigning the meta-data. In one implementation, the hardware uses the same packet metadata bits to store both the Source IP metadata and the Dest IP metadata. In another implementation, the hardware can use separate set of metadata bits to store the Source IP metadata, and another set of bits for the Dest IP metadata.

Having a SAI\_PREFIX\_COMPRESSION\_TABLE\_ATTR\_TYPE allows users to identify the supported type(s) using SAI enum capability query.

Reply:
would this work:

create prefix compression table
  - add label SIP
  - direction ingress

Create entry for table SIP
  - Add entry 10.1.1.0/24 --> meta value 10
  - Add entry 10.1.2.0/24 --> meta value 20

Create Prefix compression table
  - Add label DIP
  - Add direction Ingress

Add entries to table called DIP
 - Entry: 10.1.0.0/16 --> meta 20
 - Entry: 100.0.0.0/8 --> meta 30
 - Entry: 200.0.0.0/8 --> meata 40

Create ACL TABLE
 - Table Attribute: SAI_ACL_TABLE_ATTR_SRC_PREFIX_COMPRESSION_TABLE = SIP Table
 - Table Attribue: SAI_ACL_TABLE_ATTR_DST_PREFIX_COMPRESSION_TABLE = DIP Table
- Field support: SAI_ACL_TABLE_ATTR_FIELD_SRC_PREFIX_BINCODE
- Field support: SAI_ACL_TABLE_ATTR_FIELD_DST_PREFIX_BINCODE
- Add entry match on SAI_ACL_ENTRY_ATTR_FIELD_SRC_PREFIX_BINCODE = 10 (Add action of forward)
- Add entry match on SAI_ACL_ENTRY_ATTR_FIELD_DST_PREFIX_BINCODE = 20 (Add action of drop)

When adding the ACL entries if you don't specify a priority then it is up to the vendor on how it is implemented. If you wanted to guarantee that the drop happens before the forward then I would add a higher priority on the drop rule.

I also want to add two points:
1. When creating the Prefix Table we are using LPM so if multiple entries with the same subnet will be treated the same. In the rule you gave above for SIP you would probably want to add only one entry which is 10.1.x.x/16 and the same for the DST. 
2. in ACL these are treated differently we are not match on both SRC and DST. you are only matching on a source address or destination address in a single entry

Reply:
@rck-innovium / @helloanandhi I added a table type to table attribute. I believe this meets all the changes that were requested. Can you please let me know if you have any other requests, if not can you please approve

Reply:
There need to be is_source flag on this attribute